token-lending/sdk/src/state/reserve.rs

```diff
@@ -307,6 +313,53 @@ impl Reserve {
         })
     }
 
+    /// Calculate bonus as a percentage
+    /// the value will be in range [0, MAX_BONUS_PCT]
+    pub fn calculate_bonus(&self, obligation: &Obligation) -> Result<Decimal, ProgramError> {
+        if obligation.borrowed_value < obligation.unhealthy_borrow_value {
+            msg!("Obligation is healthy so a liquidation bonus can't be calculated");
+            return Err(LendingError::ObligationHealthy.into());
+        }
+
+        let liquidation_bonus = Decimal::from_percent(self.config.liquidation_bonus);
+        let max_liquidation_bonus = Decimal::from_percent(self.config.max_liquidation_bonus);
+        let protocol_liquidation_fee = Decimal::from_deca_bps(self.config.protocol_liquidation_fee);
+
+        // could also return the average of liquidation bonus and max liquidation bonus here, but
+        // i don't think it matters
+        if obligation.unhealthy_borrow_value == obligation.super_unhealthy_borrow_value {
+            return liquidation_bonus.try_add(protocol_liquidation_fee);
+        }
+
+        // safety:
+        // - super_unhealthy_borrow value > unhealthy borrow value because we verify
+        // the ge condition in Reserve::unpack and then verify that they're not equal from check
+        // above
+        // - borrowed_value is >= unhealthy_borrow_value bc of the check above
+        // => weight is always between 0 and 1
+        let weight = min(
+            obligation
+                .borrowed_value
+                .try_sub(obligation.unhealthy_borrow_value)?
+                .try_div(
+                    obligation
+                        .super_unhealthy_borrow_value
+                        .try_sub(obligation.unhealthy_borrow_value)?,
+                )
+                // the division above can potentially overflow if super_unhealthy_borrow_value and
+                // unhealthy_borrow_value are really close to each other. in that case, we want the
+                // weight to be one.
+                .unwrap_or_else(|_| Decimal::one()),
```

Comment:
why or_else not or? i think _or is better worse case perf.

Reply:
cargo clippy didn't like unwrap_or

Reply:
oh ok